airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,96 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Base Asynchronous Operator for kicking off a long running
+operations and polling for completion with reschedule mode.
+"""
+from functools import wraps
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.exceptions import AirflowException
+from airflow.models.xcom import XCOM_EXTERNAL_RESOURCE_ID_KEY
+
+class BaseAsyncOperator(BaseSensorOperator, SkipMixin):
+    """
+    AsyncOperators are derived from this class and inherit these attributes.
+
+    AsyncOperators must define a `submit_request` to fire a request for a
+    long running operation with a method and then executes a `poke` method
+    executing at a time interval and succeed when a criteria is met and fail
+    if and when they time out. They are effctively an opinionated way use
+    combine an Operator and a Sensor in order to kick off a long running
+    process without blocking a worker slot while waiting for the long running
+    process to complete by leveraging reschedule mode.
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+    :type mode: str
+    """
+    ui_color = '#9933ff'  # type: str
+    valid_modes = ['poke', 'reschedule']  # type: Iterable[str]
+
+    @apply_defaults
+    def __init__(self,
+                 *args,
+                 **kwargs) -> None:
+        super().__init__(mode='reschedule', *args, **kwargs)
+
+    def submit_request(self, context) -> string:
+        """
+        This method should kick off a long running operation.
+        This method should return the ID for the long running operation used
+        for polling
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+
+        :returns: a resource_id for the long running operation.
+        :rtype: str
+        """
+        raise AirflowException('Async Operators must define a `submit_request` method.')
+
+    def process_result(self, context):
+        """
+        This method can optionally be overriden to process the result of a long running operation.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+        """
+        self.log.info('Got result of {}. Done.'.format(
+                      self.get_external_resource_id(context))
+
+    def pre_execute(self, context) -> None:
```

Comment:
Is there a reason why this can't be done in execute?

Reply:
Executing code in pre_execute is dangerous because you can jam the thread.  The code executed in the execute method has a timeout.

Reply:
> Executing code in pre_execute is dangerous because you can jam the thread. The code executed in the execute method has a timeout.

Curious what you mean by this.  Are `pre_execute` and `post_execute` not meant to be subclassed?

I used them in a "WatermarkSqlOperator" for incremental loads.   Pre-execute would calculate high watermark (max date in source table) and store it as attribute.  It would also retrieve last high watermark from xcom.  Post-execute would send to xcom new high watermark after `execute` completed successfully.

I did it this way so that execute method could be overridden in subclass.

Just trying to understand what's the danger inherent in pre / post execute.